overlord/snapstate/snapmgr.go

```diff
@@ -633,6 +634,35 @@ func (m *SnapManager) ensureUbuntuCoreTransition() error {
 	return nil
 }
 
+// GenerateCookies creates snap cookies for snaps that are missing them (may be the case for snaps installed
+// before the feature of running snapctl outside of hooks was introduced, leading to a warning
+// from snap-confine).
+func (m *SnapManager) GenerateCookies(st *state.State) error {
+	var snapNames map[string]*json.RawMessage
+	if err := st.Get("snaps", &snapNames); err != nil && err != state.ErrNoState {
+		return err
+	}
+
+	var contexts map[string]string
+	if err := st.Get("snap-cookies", &contexts); err != nil {
+		if err != state.ErrNoState {
+			return fmt.Errorf("cannot get snap cookies: %v", err)
+		}
+		contexts = make(map[string]string)
+	}
+
+	for snap := range snapNames {
+		if _, ok := contexts[snap]; !ok {
+			err := m.createSnapCookie(st, snap)
+			if err != nil {
```

Comment:
(nitpick) this could be a single line: `if err := m.createSnapCookie(...; err != nil {`